File: labs/lab10_array_multiplier/array_multiplier.py
#!/usr/bin/env python3
"""Lab 10 – Array Multiplier simulation.

An N×N array multiplier computes A × B using:
  - An AND gate array to generate partial products
  - A triangle array of Full Adders to accumulate them

Run:
  python3 array_multiplier.py              # demo
  python3 array_multiplier.py <A> <B> [bits]
  Example: python3 array_multiplier.py 6 5 4
"""

import sys


# ---------------------------------------------------------------------------
# Gate primitives
# ---------------------------------------------------------------------------

def AND(a: int, b: int) -> int: return a & b

def full_adder(a: int, b: int, cin: int) -> tuple[int, int]:
    s    = a ^ b ^ cin
    cout = (a & b) | (b & cin) | (a & cin)
    return s, cout


# ---------------------------------------------------------------------------
# Array Multiplier
# ---------------------------------------------------------------------------

def array_multiplier(a: int, b: int, n: int):
    """Multiply two n-bit numbers using an array of AND gates and FAs.

    Returns (product, partial_products).
    partial_products[i][j] = bit j of partial product i (= A[j] AND B[i]).
    """
    a_bits = [(a >> i) & 1 for i in range(n)]
    b_bits = [(b >> i) & 1 for i in range(n)]

    # Step 1 – AND gate array: PP[i][j] = a_bits[j] AND b_bits[i]
    PP = [[AND(a_bits[j], b_bits[i]) for j in range(n)] for i in range(n)]

    # Step 2 – accumulate partial products row by row
    # Start with the first partial product row
    # acc[k] holds the current accumulated sum bit at position k (0 = LSB)
    # We work with 2n bits of precision.

    acc = [0] * (2 * n)

    # Initialise with PP[0] (no shifting needed, it represents A * b_bits[0])
    for j in range(n):
        acc[j] = PP[0][j]

    # Add PP[i] (shifted left by i) to accumulator
    for i in range(1, n):
        carry = 0
        for j in range(n):
            pos = i + j
            s, carry = full_adder(acc[pos], PP[i][j], carry)
            acc[pos] = s
        # propagate remaining carry
        pos = i + n
        while carry and pos < 2 * n:
            s, carry = full_adder(acc[pos], 0, carry)
            acc[pos] = s
            pos += 1

    product = sum(acc[k] << k for k in range(2 * n))
    return product, PP


# ---------------------------------------------------------------------------
# Display helpers
# ---------------------------------------------------------------------------

def fmt_bits(bits: list[int]) -> str:
    return "".join(str(b) for b in reversed(bits))


def print_multiplication(a: int, b: int, n: int):
    product, PP = array_multiplier(a, b, n)

    a_bits = [(a >> i) & 1 for i in range(n)]
    b_bits = [(b >> i) & 1 for i in range(n)]

    print(f"\n{n}-bit × {n}-bit Array Multiplier")
    print(f"  A = {fmt_bits(a_bits)}  ({a})")
    print(f"  B = {fmt_bits(b_bits)}  ({b})")
    print(f"  A × B = {a} × {b} = {product}")
    print()

    # Partial products table (long multiplication style)
    print(f"  Partial Products (PP[i] = A × B[i], shifted left by i):")
    width = 2 * n + 2
    for i in range(n):
        pp_bits = [0] * (2 * n)
        for j in range(n):
            pp_bits[i + j] = PP[i][j]
        label = f"PP[{i}] (×B[{i}]={b_bits[i]})"
        print(f"  {label:<18} = {''.join(str(pp_bits[k]) for k in range(2*n-1, -1, -1))}")

    # Result
    res_bits = [(product >> k) & 1 for k in range(2 * n)]
    print(f"  {'─' * (width + 20)}")
    print(f"  {'Product':<18} = {''.join(str(res_bits[k]) for k in range(2*n-1, -1, -1))}  ({product})")


def demo():
    print("=" * 50)
    print("Array Multiplier Demo")
    print("=" * 50)
    print_multiplication(6, 5, 4)   # 6×5=30
    print_multiplication(13, 11, 4) # 13×11=143
    print_multiplication(15, 15, 4) # 15×15=225
    print_multiplication(12, 10, 8) # 8-bit example


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    if len(sys.argv) >= 3:
        a = int(sys.argv[1])
        b = int(sys.argv[2])
        n = int(sys.argv[3]) if len(sys.argv) >= 4 else max(a, b).bit_length()
        print_multiplication(a, b, n)
    else:
        demo()


if __name__ == "__main__":
    main()

File: labs/lab10_array_multiplier/array_multiplier.cpp
// Lab 10 – Array Multiplier
// N×N unsigned multiplication via AND gate array + Full Adder triangle.
// Input: A B n
// Output: partial products + final product

#include <iostream>
#include <vector>
#include <string>
using namespace std;

pair<int,int> fullAdder(int a, int b, int cin) {
    return { a ^ b ^ cin, (a & b) | (b & cin) | (a & cin) };
}

string toBinary(long long val, int n) {
    string s(n, '0');
    for (int i = 0; i < n; i++)
        s[n-1-i] = '0' + ((val >> i) & 1);
    return s;
}

long long arrayMultiply(int A, int B, int n) {
    // AND gate array: PP[i][j] = A[j] AND B[i]
    vector<vector<int>> PP(n, vector<int>(n));
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            PP[i][j] = ((A >> j) & 1) & ((B >> i) & 1);

    // accumulate partial products into a 2n-bit array
    vector<int> acc(2*n, 0);
    for (int j = 0; j < n; j++) acc[j] = PP[0][j];

    for (int i = 1; i < n; i++) {
        int carry = 0;
        for (int j = 0; j < n; j++) {
            auto [s, co] = fullAdder(acc[i+j], PP[i][j], carry);
            acc[i+j] = s;
            carry = co;
        }
        int pos = i + n;
        while (carry && pos < 2*n) {
            auto [s, co] = fullAdder(acc[pos], 0, carry);
            acc[pos] = s; carry = co; pos++;
        }
    }

    long long result = 0;
    for (int k = 0; k < 2*n; k++) result |= (long long)acc[k] << k;
    return result;
}

int main() {
    int A, B, n;
    cout << "Enter A B n (e.g. 6 5 4): ";
    cin >> A >> B >> n;

    cout << "\n" << n << "-bit Array Multiplier\n";
    cout << "  A = " << toBinary(A, n) << "  (" << A << ")\n";
    cout << "  B = " << toBinary(B, n) << "  (" << B << ")\n\n";

    cout << "  Partial Products (PP[i] = A * B[i], shifted by i):\n";
    for (int i = 0; i < n; i++) {
        int bi = (B >> i) & 1;
        long long pp = 0;
        for (int j = 0; j < n; j++)
            pp |= (long long)(((A >> j) & 1) & bi) << (i + j);
        cout << "  PP[" << i << "] (B[" << i << "]=" << bi << ") = "
             << toBinary(pp, 2*n) << "\n";
    }

    long long product = arrayMultiply(A, B, n);
    cout << "\n  Product = " << toBinary(product, 2*n)
         << "  (" << product << ")\n";
    cout << "  Check: " << A << " x " << B << " = " << A*B
         << (product == (long long)A*B ? "  OK" : "  MISMATCH") << "\n";
    return 0;
}
